libs/common/src/change-stream/delete-many.ts

```diff
@@ -0,0 +1,72 @@
+import 'dotenv/config.js';
+import type { Filter, TransactionOptions, WithoutId } from 'mongodb';
+import { add } from 'date-fns';
+import type { AuditDetails, DeletionAuditLog, MongoDbCollectionName } from '../types';
+import { MongoDbClient } from '../mongo-db-client';
+import { generateAuditDatabaseRecordFromAuditDetails } from './generate-audit-database-record';
+import { changeStreamConfig } from './config';
+
+const { DELETION_AUDIT_LOGS_TTL_SECONDS } = changeStreamConfig;
+
+type DeleteManyParams = {
+  filter: Filter<any>;
+  collectionName: MongoDbCollectionName;
+  db: MongoDbClient;
+  auditDetails: AuditDetails;
+};
+
+const deleteManyWithAuditLogs = async ({ filter, collectionName, db, auditDetails }: DeleteManyParams) => {
+  const client = await db.getClient();
+  const session = client.startSession();
+
+  try {
+    const transactionOptions: TransactionOptions = {
+      readConcern: { level: 'snapshot' },
+      writeConcern: { w: 'majority' },
+    };
+    await session.withTransaction(async () => {
+      const collection = await db.getCollection(collectionName);
+      const documentsToDelete = await collection.find(filter, { projection: { _id: true }, session }).toArray();
+
+      if (documentsToDelete.length) {
+        const logsToInsert: WithoutId<DeletionAuditLog>[] = documentsToDelete.map(({ _id }) => ({
+          collectionName,
+          deletedDocumentId: _id,
+          auditRecord: generateAuditDatabaseRecordFromAuditDetails(auditDetails),
+          expireAt: add(new Date(), { seconds: DELETION_AUDIT_LOGS_TTL_SECONDS }),
+        }));
+
+        const deletionCollection = await db.getCollection('deletion-audit-logs');
+        const insertResult = await deletionCollection.insertMany(logsToInsert, { session });
+        if (!insertResult.acknowledged) {
+          throw new Error('Failed to create deletion audit logs');
+        }
+
+        const deleteResult = await collection.deleteMany({ $or: documentsToDelete.map(({ _id }) => ({ _id })) }, { session });
```

Comment:
I'm confused about this line, can't we just do
`const deleteResult = await collection.deleteMany(filter, { session });`

Reply:
I wanted to make sure no additional documents could creep in between the `insertMany` and `deleteMany`. Its a real edge case & I'd hope the transaction would handle it but I wanted to be sure

Reply:
Understood, in which case could we use `$in` instead of `$or` please?